inst/include/dust/dust.hpp

```diff
@@ -363,39 +461,101 @@ class Dust {
     _rng.long_jump();
   }
 
-  void set_data(std::unordered_map<size_t, std::vector<data_t>> data) {
+  void set_data(std::map<size_t, std::vector<data_t>> data) {
     _data = data;
   }
 
   std::vector<real_t> compare_data() {
     std::vector<real_t> res;
     auto d = _data.find(step());
-    // If we don't find data, we will return a vector of length 0
-    // (which we catch and convert to NULL on return to R).
     if (d != _data.end()) {
-      const size_t np = _n_pars == 0 ?
-        _particles.size() : _particles.size() / _n_pars;
       res.resize(_particles.size());
+      compare_data(res, d->second);
+    }
+    return res;
+  }
+
+  void compare_data(std::vector<real_t>& res, const std::vector<data_t>& data) {
+    const size_t np = _particles.size() / n_pars_effective();
 #ifdef _OPENMP
-      #pragma omp parallel for schedule(static) num_threads(_n_threads)
+    #pragma omp parallel for schedule(static) num_threads(_n_threads)
 #endif
-      for (size_t i = 0; i < _particles.size(); ++i) {
-        res[i] = _particles[i].compare_data(d->second[i / np], _rng.state(i));
+    for (size_t i = 0; i < _particles.size(); ++i) {
+      res[i] = _particles[i].compare_data(data[i / np], _rng.state(i));
+    }
+  }
+
+  std::vector<real_t> filter(bool save_history) {
+    if (_data.size() == 0) {
+      throw std::invalid_argument("Data has not been set for this object");
+    }
+
+    const size_t n_particles = _particles.size();
+    const size_t n_particles_each = n_particles / n_pars_effective();
+    std::vector<real_t> log_likelihood(n_pars_effective());
+    std::vector<real_t> log_likelihood_step(n_pars_effective());
+    std::vector<real_t> weights(n_particles);
+    std::vector<size_t> kappa(n_particles);
+
+    if (save_history) {
+      filter_state_.resize(_index.size(), _particles.size(), _data.size());
+      state(filter_state_.history_value_iterator());
+      filter_state_.advance();
+    }
+
+    for (auto & d : _data) {
```

Comment:
Ok, this iteration motivates the change to a map. The key will be positive integers. Remind me, does it always go up in increments of 1; does it always start from 0? Or does the usage here totally relax this?

An alternative would be two std::vectors, one of data_step, one of std::vector<data_t>. This is what I think we should/will have to use in the CUDA (complications about user defined data_t struct aside). Maybe we should switch to that here, it would change the use of find in `compare_data` above a bit, I think it ends up being very similar to this solution

Reply:
goes up in arbitrary increments, but always increasing. Can we hold off on a change like that until this bit starts getting moved to the device as it would not be clearer or faster at the moment